service/s3/s3crypto/aes_gcm.go

```diff
@@ -0,0 +1,105 @@
+package s3crypto
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"io"
+	"io/ioutil"
+)
+
+// AESGCM Symmetric encryption algorithm. Since Golang designed this
+// with only TLS in mind. We have to load it all into memory meaning
+// this isn't streamed.
+type AESGCM struct {
+	aead  cipher.AEAD
+	nonce []byte
+}
+
+// NewAESGCM creates a new AES GCM cipher. Expects keys to be of
+// the correct size.
+//
+// Example:
+//
+//	cd := &s3crypto.CipherData{
+//		Key: key,
+//		"IV": iv,
+//	}
+//	cipher, _ := s3crypto.NewAESGCM(cd)
+func NewAESGCM(cd CipherData) (Cipher, error) {
+	block, err := aes.NewCipher(cd.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AESGCM{aesgcm, cd.IV}, nil
+}
+
+// Encrypt will encrypt the data using AES GCM
+// Tag will be included as the last 16 bytes of the slice
+func (c *AESGCM) Encrypt(src io.Reader) io.Reader {
+	reader := &gcmEncryptReader{
+		encrypter: c.aead,
+		nonce:     c.nonce,
+		src:       src,
+	}
+	return reader
+}
+
+type gcmEncryptReader struct {
+	encrypter cipher.AEAD
+	nonce     []byte
+	src       io.Reader
+	buf       *bytes.Buffer
+}
+
+func (reader *gcmEncryptReader) Read(data []byte) (int, error) {
+	if reader.buf == nil {
+		b, err := ioutil.ReadAll(reader.src)
+		if err != nil {
+			return len(b), err
+		}
+		b = reader.encrypter.Seal(nil, reader.nonce, b, nil)
+		reader.buf = bytes.NewBuffer(b)
```

Comment:
could the first param be `b[:0]` so that the byte array will be reused for output? If seal needs to grow the array it can, and the returned `[]byte`  (b) will be updated as needed. But gives Seal a chance to reuse the byte array.

https://golang.org/pkg/crypto/cipher/#AEAD

```go
b = reader.encrypter.Seal(b[:0], reader.nonce, b, nil)
```